posts: add Adding a Timeout feature to Rustgem blogpost (#4)
mairacanal/mairacanal.github.io: _posts/2023-03-11-Adding-a-Timeout-feature-to-Rustgem.md

---
title: "Adding a Timeout feature to Rustgem"
date: 2023-03-11
author: Maíra Canal
layout: post
permalink: /adding-timeout-rustgem/
categories: Genel
tags: [igalia, graphics]
---

> After my last blogpost, I kept developing the Rust for VGEM driver, also known

Reviewer: The sentence is a bit unclear. Maybe you mean this?

```suggestion
After my last blogpost, I kept developing the Rust VGEM driver, also known
```

> as rustgem, for now. Previously, I have developed two important features of the

Reviewer: Unclear if the "for now" referes to the name being temporary, in which case I'd remove the comma, or if you're refering to your current general plans. In that case, I'd remove "for now".

```suggestion
as rustgem for now. Previously, I had developed two important features of the
```

driver: the ability to attach a fence and the ability to signal a fence. But,
one important feature is still missing: the ability to prevent hangs. Currently,
if the fence is not signaled, the driver will simply hang. So, we can create a
callback that signals the fence when the fence is not signaled by the user for
more than 10 seconds.

> In order to create this callback, we need to have a Timer that will trigger the
> callback when the fence is not signaled by the user for more than 10 seconds.

Reviewer: A bit redundant with the end of the previous paragraph. How about...

```suggestion
In order to create this callback, we need to have a Timer that will trigger it after the specified amount of time.
```

Gladly, the Linux kernel provides us with a Timer that can be set with a
callback and a timeout. But, to use it in the Rust code, we need to have a safe
abstraction, that will ensure that the code is safe under some assumptions.

# First Attempt: writing a Timer abstraction
---

So, initially, I was developing an abstraction on my own as I checked the RfL
tree and there were no Timer abstractions available.

The most important question here is "how can we guarantee access to other

> objects inside the callback?". The callback only has, as parameters, a pointer
> to the `struct timer_list`. Naturally, we can think about using a `container_of`

Reviewer: ```suggestion
objects inside the callback?". The callback only receives a pointer
to the `struct timer_list` as its single argument. Naturally, we can think about using a `container_of`
```

macro. So, in order to make the compatibility layer between Rust and the C
callback, I decided to store the object inside the Timer. Yep, I didn't like

> that a lot, but it was the solution I came in at the time. So, the struct looked

Reviewer: ```suggestion
that a lot, but it was the solution I came up with at the time. So, the struct looked
```

    /// Type of the Inner data inside the Timer
    type Inner;

    /// Timer callback
    fn timer_callback(timer: &UniqueTimer<Self, Self::Inner>);
}
```

With those two questions solved, it seems that we are all set and good to go.
So, we can create methods to initialize the timer and modify the timer's
timeout, implement the Drop trait, and use the following callback by default:

```rust
unsafe extern "C" fn timer_callback<T: TimerOps<Inner = D>, D: Sized>(
    timer: *mut bindings::timer_list,
) {
    let timer = crate::container_of!(timer, UniqueTimer<T, D>, timer)
			    as *mut UniqueTimer<T, D>;

    // SAFETY: The caller is responsible for passing a valid timer_list subtype
    T::timer_callback(unsafe { &mut *timer });
}
```

All should work, right? Well... No, I didn't really mention how I was allocating
memory. And let's say I was initially allocating it wrongly and therefore, the
`container_of` macro was pointing to the wrong memory space.

Initially, I was allocating only `timer` with the kernel memory allocator
`krealloc` and allocating the rest of the struct with Rust's memory allocator.
By making such a mess, `container_of` wasn't able to point to the right

> direction.

Reviewer: I found this a bit confusing. "Direction" is the memory address? Or is the word used in a general meaning?

So, I had to change things a bit to allocate the whole struct `UniqueTimer` with
the kernel's memory allocator. But, `krealloc` returns a raw pointer and it
would be nice for the final user to get a raw pointer to the object. So, I
wrapped up inside another struct that could be dereferenced into the
`UniqueTimer` object.

```rust
/// A generic Timer Object
///
/// This object should be instantiated by the end user, as it holds

> /// a unique reference to the struct UniqueTimer. The UniqueTimer

Reviewer: ```suggestion
/// a unique reference to the UniqueTimer struct. The UniqueTimer
```

/// methods can be used through it.
pub struct Timer<T: TimerOps<Inner = D>, D>(*mut UniqueTimer<T, D>);

impl<T: TimerOps<Inner = D>, D> Timer<T, D> {
    /// Create a timer for its first use
    pub fn setup(inner: D) -> Self {
        let t = unsafe {
            bindings::krealloc(
                core::ptr::null_mut(),
                core::mem::size_of::<UniqueTimer<T, D>>(),
                bindings::GFP_KERNEL | bindings::__GFP_ZERO,
            ) as *mut UniqueTimer<T, D>
        };

        // SAFETY: The pointer is valid, so pointers to members are too.
        // After this, all fields are initialized.
        unsafe {
            addr_of_mut!((*t).inner).write(inner);
            bindings::timer_setup(addr_of_mut!((*t).timer), Some(timer_callback::<T, D>), 0)
        };

        Self(t)
    }
}
```

> And them, the `container_of` macro is now working! Now, I could setup a Timer

Reviewer: ```suggestion
And then the `container_of` macro started working! Now, I could setup a Timer
```

for each fence and keep the fence inside the timer. Finally, I could use the
fence inside the timer to signal it when it was not signaled by the user for
more than 10 seconds.

```rust
impl TimerOps for VgemFenceOps {
    type Inner = UniqueFence<Self>;

    fn timer_callback(timer: &UniqueTimer<Self, UniqueFence<Self>>) {
        let _ = timer.inner().signal();
    }
}
```

So, I tested the driver with IGT using the `vgem_slow` test and it was now
passing! All IGT tests were passing and it looked like the driver was

> practically completed (a part from some FIXME problems). But, let's see if this

Reviewer: ```suggestion
practically completed (some FIXME problems notwithstanding). But, let's see if this
```

`u32` value. Initially, everything looks fine, the pointer field points to the
value field in memory address A, which contains a valid `u32`, and all pointers
are **valid**. But Rust has the freedom to move values around memory. For
example, if we pass this struct into another function, it might get moved to a
different memory address. So, the once valid pointer is no longer valid, because
when we move the struct, the struct's fields change their address, but not their
value. So, the pointer fields still point to the memory address A, although the
value field is located at the memory address B now. This is really bad and can
lead to UB.

The solution is to make `timer_list` implement the `!Unpin` trait. This means
that to use this type safely, we can't use regular pointers for self-reference.
Instead, we use special pointers that "pin" their values into place, ensuring
they can't be moved.

Still looking at the `struct timer_list`, it is possible to notice that a timer
can queue itself in the timer function. This functionality is not covered by my
current abstraction.

Moreover, I was using jiffies to modify the timeout duration and I was adding a

> `Duration` to the jiffies. This is problematic, because this can cause a data
> race. So, the reading of jiffies should be atomic and adding a Duration to
> jiffies should be a atomic operation.

Reviewer: ```suggestion
`Duration` to the jiffies. This is problematic, because it can cause data
races. Reading jiffies and adding a duration to them should be an atomic operation.
```

> Huge thanks to the RfL folks that pointed the errors in my implementation!

> With all this problems pointed, it is time to fix it! I could have reimplemented

Reviewer: ```suggestion
With all these problems pointed out, it is time to fix them! I could have reimplemented
```
